sys/uartpb/include/uartpb/UartProtobuf.hpp

```diff
@@ -0,0 +1,259 @@
+
+    uint8_t msg_length_bytes_number_ = 4;           ///< number of bytes needed to encode message length
+    uint8_t msg_length_bytes_received_ = 0;         ///< number of bytes received for current message length
+    uint32_t msg_length_ = 0;                       ///< message length
+    uint32_t msg_bytes_received_ = 0;               ///< number of bytes received for current message
+    uint8_t input_buffer_[INPUT_MESSAGE_LENGTH_MAX];      ///< buffer used to decode incoming messages
+    uint8_t output_buffer_[OUTPUT_MESSAGE_LENGTH_MAX];    ///< buffer used to encode outgoing messages
+};
+
+template <typename InputMessageType, typename OutputMessageType,
+          void (*message_handler)(const InputMessageType &)>
+UartProtobuf<InputMessageType, OutputMessageType, message_handler>::
+UartProtobuf(
+    const pb_msgdesc_t *input_message_fields,
+    const pb_msgdesc_t *output_message_fields,
+    uart_t uart_dev, uint32_t uart_speed) :
+    input_message_fields_(input_message_fields),
+    output_message_fields_(output_message_fields),
+    uart_dev_(uart_dev), uart_speed_(uart_speed)
+{
+    ringbuffer_init(&rx_buf_, rx_mem_, INPUT_MESSAGE_LENGTH_MAX);
+}
+
+template <typename InputMessageType, typename OutputMessageType,
+          void (*message_handler)(const InputMessageType &)>
+bool UartProtobuf<InputMessageType, OutputMessageType, message_handler>::
+connect()
+{
+    int res = uart_init(uart_dev_, uart_speed_, uart_rx_cb_wrapper, (void *)this);
+    return res == UART_OK;
+}
+
+template <typename InputMessageType, typename OutputMessageType,
+          void (*message_handler)(const InputMessageType &)>
+void UartProtobuf<InputMessageType, OutputMessageType, message_handler>::
+start_reader()
+{
+    reader_pid_ = thread_create(
+        reader_stack_, sizeof(reader_stack_), UARTPB_READER_PRIO,
+        0, message_reader_wrapper, (void *)this, "reader");
+}
+
+template <typename InputMessageType, typename OutputMessageType,
+          void (*message_handler)(const InputMessageType &)>
+void UartProtobuf<InputMessageType, OutputMessageType, message_handler>::
+uart_rx_cb(uint8_t data)
+{
+    if (msg_length_bytes_received_ < msg_length_bytes_number_) {
+        ((uint8_t *)&msg_length_)[msg_length_bytes_received_++] = data;
+        msg_bytes_received_ = 0;
+        return;
+    }
+
+    if (msg_bytes_received_ < msg_length_) {
+        ringbuffer_add_one(&rx_buf_, data);
+        msg_bytes_received_++;
+    }
+
+    if (msg_bytes_received_ == msg_length_) {
+        msg_t msg;
+        msg.content.value = msg_length_;
+        msg_send(&msg, reader_pid_);
+        msg_length_bytes_received_ = 0;
+    }
+}
+
+template <typename InputMessageType, typename OutputMessageType,
+          void (*message_handler)(const InputMessageType &)>
+void UartProtobuf<InputMessageType, OutputMessageType, message_handler>::
+decode_message(uint32_t message_length, uint8_t *buffer)
+{
+    InputMessageType message;
```

Comment:
Noob question: as the template InputMessageType seems to be used only there. Could the template be defined for this function and not for the class ?